AY2122S1-CS2103T-T17-3/tp
# Reimplement tag

### Description



A revamp of the implementation of `Tag` implementation in preparation for the TagListPanel feature to be added in my next PR.

### The new `Model` architecture as a result of this: 
![image](https://user-images.githubusercontent.com/24221801/137860968-e7ffb537-06aa-4053-9b9e-c36c5c8a65f5.png)

There is a `Tag` list in the `AddressBook`, which `Person` references. This allows `AddressBook` to only require one `Tag` object per unique tag, instead of each `Person` needing their own `Tag` objects.


### Type of change


### How to test



No new features are introduced in this PR so you can just check that the tag `t/` works correctly as usual.

Some example commands you could try out: 

Command | Expected Behaviour 
--- | --- 
edit 0 t/family | The tags under client of id 0 should update to `family` 
edit 0 t/family t/friends | The tags under client of id 0 should update to `family` and `friends` 
add n/John Doe e/johnd@example.com t/assassin | The added client should have a tag `assassin` associated with him

### Checklist
